src/org/labkey/test/util/Crawler.java

```diff
@@ -1039,27 +1044,31 @@ private List<UrlToCheck> crawlLink(final UrlToCheck urlToCheck)
                                 }
                             }
                         }
-                        List<String> linkAddresses = linksWithAttributes.stream().map(Pair::getLeft).collect(Collectors.toList());
-                        List<String> formAddresses = _test.getFormAddresses();
-                        linkAddresses.addAll(formAddresses);
 
-                        for (String url : linkAddresses)
+                        int nextDepth = depth + 1;
+
+                        for (Pair<String, ?> p : linksWithAttributes)
                         {
+                            String url = p.getLeft();
                             try
                             {
-                                UrlToCheck candidateUrl = new UrlToCheck(actualUrl, url, depth + 1);
-                                if (candidateUrl.isVisitableURL())
-                                {
-                                    candidateUrl.setFromForm(formAddresses.contains(url));
-                                    newUrlsToCheck.add(candidateUrl);
-                                }
+                                newUrlsToCheck.add(new UrlToCheck(actualUrl, url, nextDepth));
                             }
                             catch (IllegalArgumentException badUrl)
                             {
-                                if (!formAddresses.contains(url)) // forms might have strange target action (e.g. '../formulations')
-                                {
-                                    throw new AssertionError("Unable to parse link: \"" + url + "\". " + originMessage, badUrl);
-                                }
+                                throw new IllegalArgumentException("Unable to parse link: \"" + url + "\". " + originMessage, badUrl);
+                            }
+                        }
+
+                        for (String url : _test.getFormAddresses())
+                        {
+                            try
+                            {
+                                newUrlsToCheck.add(new UrlToCheck(actualUrl, url, nextDepth).setFromForm(true));
+                            }
+                            catch (IllegalArgumentException ignore)
+                            {
+                                // forms might have strange target action (e.g. '../formulations')
```

Comment:
I'm not sure I understand the code entirely but it looks like the constructor to UrlToCheck will throw an IllegalArgumentException if the depth is negative. Are you ok with eating that exception here?

Reply:
That shouldn't be a problem. The crawler used to have a way to indicate a `-1` depth in order to crawl a single link (I forget the specifics). I replaced that with the `validatePage` method at some point. There is not, currently, any way for the depth to be `-1`.